packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -173,12 +482,632 @@ describe('ReactMount', () => {
+        '    <div data-ssr-mismatch-padding-before="1" />\n' +
+        '    <div data-ssr-mismatch-padding-before="2" />\n' +
+        '    <div data-ssr-mismatch-padding-before="3" />\n' +
+        '    <div data-ssr-mismatch-padding-before="4" />\n' +
+        '    <div data-ssr-mismatch-padding-before="5" />\n' +
+        '    <div data-ssr-mismatch-padding-before="6" />\n' +
+        '    <div data-ssr-mismatch-padding-before="7" />\n' +
+        '    <div data-ssr-mismatch-padding-before="8" />\n' +
+        '    <div data-ssr-mismatch-padding-before="9" />\n' +
+        '    <div data-ssr-mismatch-padding-before="10" />\n' +
+        '    <div data-ssr-mismatch-padding-before="11" />\n' +
+        '    <div data-ssr-mismatch-padding-before="12" />\n' +
+        "+   {'SSRMismatchTest client text'}\n" +
+        '  </div>\n\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate inserts an element after a comment node (insertion diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment --></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment -->\n' +
+        '+   <span />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate replaces an element after a comment node (replacement diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment <-- --&gt; > --><div></div></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+          <div />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment <-- --&gt; > -->\n' +
+        '-   <div />\n' +
+        '+   <span />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate inserts an element after a non-typical node (insertion diff)', () => {
+    // A non-typical node is a node that is not typically seen in the DOM: not an element, text, or comment.
+    // This is an artificial test case to check how we print non-typical nodes if they somehow end up in the DOM.
+    const xml = document.createElement('xml');
+    xml.appendChild(
+      document.createProcessingInstruction(
+        'dom-processing-instruction',
+        'content > ',
+      ),
+    );
+
+    expect(() => ReactDOM.hydrate(<div />, xml)).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <div> in <xml>.\n\n' +
+        '  <xml>\n' +
+        '    <?dom-processing-instruction content &gt; ?>\n' +
```

Comment:
Same as for HTML comments, but for other non-hydratable nodes, escape the `>` with `&gt;` in the diff.